### PR TITLE
Make docker app work with rrdcached

### PR DESCRIPTION
**Pull request by djamp42**
Okay opening up another pull request on the docker over rrdcached issue, i have moved the methods in the \LibreNMS\Store\Rrd class and tested with both rrdcached and without and both load docker app just fine now. The only thing that is still a question for me is the sync vs async for sending the command to rrdtool. I forced it to sync from what the commend said. 

If this is approved i can open another pull requests and fix all the other apps that are broken with rrdcached. 

Prior pull requests here
https://github.com/librenms/librenms/pull/12704
https://github.com/librenms/librenms/pull/12689

DO NOT DELETE THE UNDERLYING TEXT

#### Please note

> Please read this information carefully. You can run `./lnms dev:check` to check your code before submitting.

- [X] Have you followed our [code guidelines?](https://docs.librenms.org/Developing/Code-Guidelines/)
- [X] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

### CODE REVIEW

File: LibreNMS/Data/Store/Rrd.php [hunk 1]
```diff
@@ -442,6 +444,65 @@ public function buildCommand($command, $filename, $options)
...
+            $pattern = sprintf('%s-%s-%s-%s', 'app', $app_name, $app_id, $category);
+        } else {
+            $pattern = sprintf('%s-%s-%s', 'app', $app_name, $app_id);
+        }
+
+        foreach ($rrdfilearray as $rrd) {
+            
+            if (str_contains($rrd, $pattern)) {
+                $filename = basename($rrd, '.rrd');
+                $entry = explode($separator, $filename, 4 + $offset)[3 + $offset];
```
Comment (Jellyfrog) [inline LibreNMS/Data/Store/Rrd.php:496]: Offset doesn't exist

File: LibreNMS/Data/Store/Rrd.php [hunk 2]
```diff
@@ -384,6 +384,8 @@ private function command($command, $filename, $options)
         if ($command == 'last' && $this->init(false)) {
             // send this to our synchronous process so output is guaranteed
             $output = $this->sync_process->sendCommand($cmd);
+        } elseif ($command == 'list') {
```
Comment (murrant) [inline LibreNMS/Data/Store/Rrd.php:387]: This should be combined with the above if statement.
```php
if (in_array($command, ['last', 'list']) && $this->init(false)) {
```
The second statement makes sure the sync process is started.

File: LibreNMS/Data/Store/Rrd.php [hunk 3]
```diff
@@ -442,6 +444,67 @@ public function buildCommand($command, $filename, $options)
...
+     */
+    public function getRrdFiles($device)
+    {
+        if ($this->rrdcached) {
+            $filename = sprintf('/%s', $device['hostname']);
+            $rrd_files = $this->command('list', $filename, '');
+            // Command output is an array, create new array with each filename as a item in array.
+            $rrdfilearray = preg_split('/\s+/', trim($rrd_files[0]));
+        } else {
+            $rrddir = self::dirFromHost($device['hostname']);
```
Comment (murrant) [inline LibreNMS/Data/Store/Rrd.php:462]: Should be `$rrddir = $this->dirFromHost($device['hostname']);`

Rrd is a singleton that stays loaded in memory after it is started.   The external alias \Rrd is a static way to access that singleton.

File: LibreNMS/Data/Store/Rrd.php [hunk 4]
```diff
@@ -442,6 +444,67 @@ public function buildCommand($command, $filename, $options)
...
+     * @param string  $app_name name of app to be searched
+     * @param string  $category which category of graphs are searched
+     * @return array $rrdfilearray array of rrd files for this host
+     */
+    public function getRrdApplicationArrays($device, $app_id, $app_name, $category = null)
+    {
+        $entries = [];
+        $separator = '-';
+
+        $rrdfilearray = self::getRrdFiles($device);
```
Comment (murrant) [inline LibreNMS/Data/Store/Rrd.php:484]: `$rrdfilearray = $this->getRrdFiles($device);`

File: LibreNMS/Data/Store/Rrd.php [hunk 5]
```diff
@@ -442,6 +442,67 @@ public function buildCommand($command, $filename, $options)
...
+     * @param array $device device for which we get the rrd's
+     * @return array $rrdfilearray array of rrd files for this host
+     */
+    public function getRrdFiles($device)
+    {
+        if ($this->rrdcached) {
+            $filename = sprintf('/%s', $device['hostname']);
+            $rrd_files = $this->command('list', $filename, '');
+            // Command output is an array, create new array with each filename as a item in array.
+            $rrdfilearray = preg_split('/\s+/', trim($rrd_files[0]));
```
Comment (Jellyfrog) [inline LibreNMS/Data/Store/Rrd.php:458]:
```suggestion
            $rrdfilearray = explode("\n", trim($rrd_files[0]));
```

Comment (Jellyfrog) [reply LibreNMS/Data/Store/Rrd.php:458]: The last line will be  `"OK u:0.01 s:0.01 r:0.02"` , maybe we should drop it?

**Comment by Jellyfrog** [comment]
I added sorting (so both ways return the result in the same way)

**Review by murrant — APPROVED** [review]

**Comment by f0o** [comment]
@Jellyfrog @murrant Just noticed that this (specifically https://github.com/librenms/librenms/pull/12746/files#diff-fb57ae1ea0f72f33f98641088606fd88c0f3b1f774eccd2828af80d39ee50803R456) fails with rrdcached 1.4.8 - Unsure when `list` operation was introduced but seeing how rrdcached 1.4.x is supported and https://github.com/librenms/librenms/blob/master/LibreNMS/Data/Store/Rrd.php#L68 defaults to 1.4x it might befenit from a strict version check

**Comment by murrant** [comment]
@f0o not sure what to do about it, since we don't know when rrdlist was introduced...

rrdtool version is detected at runtime now, so it should never fall back to 1.4

**Comment by f0o** [comment]
@murrant what I mean is that anything calling `getRrdApplicationArrays` will return empty array if the installation uses rrdcached and is not on a very recent version. It might be fine with actually removing that whole rrdcached thing unless somebody knows which version introduces `list`. I can say 1.4.x doesnt but 1.7.x does.

**Comment by murrant** [comment]
Aren't all supported os on 1.7+?

**Pull request merged**